File: carve_helpers.py
"""Dual-projection voxel carving helpers."""

from typing import Optional

import numpy as np


def close_side_z_gaps(side: np.ndarray, max_gap: int = 3) -> np.ndarray:
    """
    Per X column, fill short gaps along Z in the side silhouette (side view holes).
    """
    if max_gap <= 0:
        return side
    side = side.copy()
    size = side.shape[0]
    for x in range(size):
        zs = np.where(side[:, x])[0]
        if zs.size < 2:
            continue
        for i in range(len(zs) - 1):
            a, b = int(zs[i]), int(zs[i + 1])
            if 1 < b - a <= max_gap + 1:
                side[a : b + 1, x] = True
    return side


def z_at_column(side: np.ndarray, x: int) -> np.ndarray:
    return np.where(side[:, x])[0]


def y_at_column(front: np.ndarray, x: int) -> np.ndarray:
    return np.where(front[:, x])[0]


def pick_z_for_front(
    y: int,
    x: int,
    side: np.ndarray,
    size: int,
    depth_face_bridge: bool = True,
) -> Optional[int]:
    """
    Choose Z for voxel (y, x, z).

    When side view has depth gaps at column x, use the two boundary faces
    (z_min and z_max of side foreground) so front projection stays complete
  without filling the entire depth slab (preserves side view holes).
    """
    strict = z_at_column(side, x)
    if strict.size == 0:
        return None

    if depth_face_bridge and strict.size >= 1:
        z_lo, z_hi = int(strict[0]), int(strict[-1])
        if z_hi > z_lo + 1:
            pool = np.array([z_lo, z_hi], dtype=int)
            return int(pool[(y * len(pool)) // max(1, size) % len(pool)])

    return int(strict[(y * strict.size) // max(1, size) % strict.size])


def pick_y_for_side(
    z: int,
    x: int,
    front: np.ndarray,
    size: int,
    depth_face_bridge: bool = True,
) -> Optional[int]:
    """Symmetric: use front column Y extent faces when bridging."""
    strict = y_at_column(front, x)
    if strict.size == 0:
        return None

    if depth_face_bridge and strict.size >= 1:
        y_lo, y_hi = int(strict[0]), int(strict[-1])
        if y_hi > y_lo + 1:
            pool = np.array([y_lo, y_hi], dtype=int)
            return int(pool[(z * len(pool)) // max(1, size) % len(pool)])

    return int(strict[(z * strict.size) // max(1, size) % strict.size])


def carve_dual_cover(
    front: np.ndarray,
    side: np.ndarray,
    *,
    depth_face_bridge: bool = True,
) -> np.ndarray:
    """
    Build voxels: OR_z == front, OR_y == side (exact on side mask pixels).
    depth_face_bridge: place voxels at depth/Y faces when the other view has gaps.
    """
    size = front.shape[0]
    voxels = front[:, :, np.newaxis] & side.T[np.newaxis, :, :]

    for y in range(size):
        for x in range(size):
            if not front[y, x] or np.any(voxels[y, x, :]):
                continue
            z = pick_z_for_front(y, x, side, size, depth_face_bridge)
            if z is not None:
                voxels[y, x, z] = True

    for z in range(size):
        for x in range(size):
            if not side[z, x] or np.any(voxels[:, x, z]):
                continue
            y = pick_y_for_side(z, x, front, size, depth_face_bridge)
            if y is not None:
                voxels[y, x, z] = True

    return voxels


def projection_front(voxels: np.ndarray) -> np.ndarray:
    return np.any(voxels, axis=2)


def projection_side(voxels: np.ndarray) -> np.ndarray:
    return np.any(voxels, axis=0).T

File: static/js/carve_helpers.js
/** Dual-projection carving with depth face bridging. */

export function closeSideZGaps(side, size, maxGap) {
  if (maxGap <= 0) return side.slice();
  const out = side.slice();
  for (let x = 0; x < size; x++) {
    const zs = [];
    for (let z = 0; z < size; z++) {
      if (out[z * size + x]) zs.push(z);
    }
    for (let i = 0; i < zs.length - 1; i++) {
      const a = zs[i];
      const b = zs[i + 1];
      if (b - a > 1 && b - a <= maxGap + 1) {
        for (let z = a; z <= b; z++) out[z * size + x] = 1;
      }
    }
  }
  return out;
}

export function zAtColumn(side, size, x) {
  const zs = [];
  for (let z = 0; z < size; z++) {
    if (side[z * size + x]) zs.push(z);
  }
  return zs;
}

export function yAtColumn(front, size, x) {
  const ys = [];
  for (let y = 0; y < size; y++) {
    if (front[y * size + x]) ys.push(y);
  }
  return ys;
}

export function pickZForFront(y, x, side, size, depthFaceBridge) {
  const strict = zAtColumn(side, size, x);
  if (!strict.length) return null;
  if (depthFaceBridge) {
    const zLo = strict[0];
    const zHi = strict[strict.length - 1];
    if (zHi > zLo + 1) {
      const pool = [zLo, zHi];
      return pool[(y * pool.length) / size % pool.length | 0];
    }
  }
  return strict[(y * strict.length) / size % strict.length | 0];
}

export function pickYForSide(z, x, front, size, depthFaceBridge) {
  const strict = yAtColumn(front, size, x);
  if (!strict.length) return null;
  if (depthFaceBridge) {
    const yLo = strict[0];
    const yHi = strict[strict.length - 1];
    if (yHi > yLo + 1) {
      const pool = [yLo, yHi];
      return pool[(z * pool.length) / size % pool.length | 0];
    }
  }
  return strict[(z * strict.length) / size % strict.length | 0];
}

export function zPool(side, size, x, depthFaceBridge) {
  const strict = zAtColumn(side, size, x);
  if (!strict.length) return strict;
  const zLo = strict[0];
  const zHi = strict[strict.length - 1];
  if (depthFaceBridge && zHi > zLo + 1) {
    const set = new Set(strict);
    set.add(zLo);
    set.add(zHi);
    return [...set].sort((a, b) => a - b);
  }
  return strict;
}

export function carveDualCover(front, side, size, depthFaceBridge = true) {
  const voxels = new Uint8Array(size * size * size);
  const idx = (y, x, z) => y * size * size + x * size + z;

  for (let y = 0; y < size; y++) {
    for (let x = 0; x < size; x++) {
      if (!front[y * size + x]) continue;
      for (let z = 0; z < size; z++) {
        if (side[z * size + x]) voxels[idx(y, x, z)] = 1;
      }
    }
  }

  for (let y = 0; y < size; y++) {
    for (let x = 0; x < size; x++) {
      if (!front[y * size + x]) continue;
      let any = false;
      for (let z = 0; z < size; z++) {
        if (voxels[idx(y, x, z)]) {
          any = true;
          break;
        }
      }
      if (any) continue;
      const z = pickZForFront(y, x, side, size, depthFaceBridge);
      if (z != null) voxels[idx(y, x, z)] = 1;
    }
  }

  for (let z = 0; z < size; z++) {
    for (let x = 0; x < size; x++) {
      if (!side[z * size + x]) continue;
      let any = false;
      for (let y = 0; y < size; y++) {
        if (voxels[idx(y, x, z)]) {
          any = true;
          break;
        }
      }
      if (any) continue;
      const y = pickYForSide(z, x, front, size, depthFaceBridge);
      if (y != null) voxels[idx(y, x, z)] = 1;
    }
  }

  return voxels;
}
